### DISCUSSION TITLE
NEAR vs Soroban: first impressions: holy contract sizes, Batman!

### DISCUSSION THREAD

**Discussion by chadoh** [stellar/sorobanathon#19] [General]
The code in [this NEAR codebase](https://github.com/AhaLabs/hello-world-soroban-comparison) has similar functionality to [Soroban's quick start contract](https://soroban.stellar.org/docs/getting-started/quick-start), so I could compare contract sizes using NEAR's suggested defaults vs Soroban's.

Here's the Soroban code:

```rust
#![no_std]
use soroban_sdk::{contractimpl, symbol, vec, Env, Symbol, Vec};

pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn hello(env: Env, to: Symbol) -> Vec<Symbol> {
        vec![&env, symbol!("Hello"), to]
    }
}

#[cfg(test)]
mod test {
    use super::{Contract, ContractClient};
    use soroban_sdk::{symbol, vec, BytesN, Env};

    #[test]
    fn test() {
        let env = Env::default();
        let contract_id = BytesN::from_array(&env, &[0; 32]);
        env.register_contract(&contract_id, Contract);
        let client = ContractClient::new(&env, &contract_id);

        let words = client.hello(&symbol!("Dev"));
        assert_eq!(
            words,
            vec![&env, symbol!("Hello"), symbol!("Dev"),]
        );
    }
}
```

Here's the NEAR code:

```rust
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
use near_sdk::{near_bindgen, AccountId};

#[near_bindgen]
#[derive(Default, BorshDeserialize, BorshSerialize)]
pub struct Counter;

#[near_bindgen]
impl Counter {
    pub fn hello(&self, to: &AccountId) -> (String, AccountId) {
        (String::from("hello"), to.clone())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hello() {
        // instantiate a contract variable with the counter at zero
        let contract = Counter;
        let account_id: AccountId = "near".parse().unwrap();
        assert_eq!(
            ("hello".to_string(), account_id.clone()),
            contract.hello(&account_id)
        );
    }
}
```

# The Results: Holy contract size, Batman!

Soroban's contract size:

<img width="1081" alt="soroban contract size: 3.3 kilobytes!" src="https://user-images.githubusercontent.com/221614/193348126-cc5c5221-0afd-47b3-9e34-cc9d5a98b15b.png">

NEAR's contract size:

<img width="1049" alt="near contract size: 99 kilobytes!" src="https://user-images.githubusercontent.com/221614/193348205-217c1a83-dc87-4470-ac24-f4ec3db3e3b3.png">

Soroban contracts are 3% the size!

NEAR smart contracts built with Rust are big for two main reasons:

- `std`, the standard library, which Soroban requires you skip (see that `#![no_std]`?)
- JSON serialization for the function arguments, which uses [serde_json](https://docs.rs/serde_json/latest/serde_json/)

I know there are ways to avoid both of these with NEAR smart contracts, but the only people I've known to attempt it are Rust experts. Soroban seems like they're trying to give all developers this superpower by default.

# But it doesn't end there!

With just some [build process changes](https://soroban.stellar.org/docs/tutorials/build-optimized), you can build an _optimized_ version of your Soroban smart contract:

<img width="1186" alt="soroban contract optimized: 477 bytes!" src="https://user-images.githubusercontent.com/221614/193348778-b008f5f2-5ce9-40e0-b381-b7eb4de70e49.png">

3.3 kilobytes wasn't small enough! Had to make it about 10x smaller yet.

Damn, y'all. Settle down.

# Is there any future for Stellar, though?

I met [someone](https://twitter.com/tyvdh) from the Stellar Developer Relations team while at [NEARCON](https://nearcon.org/). When I mentioned Stellar to a NEAR maxi, he said something like "Stellar's not going anywhere, though."

I mean, maybe? 

Stuff I like about Stellar:

- long track record (started in 2013, before Ethereum)
- focused on safety, scalability and global financial inclusion
- have done lots of ([sloppy](https://www.coindesk.com/business/2019/12/13/stellar-tried-to-give-away-2b-xlm-tokens-on-keybase-then-the-spammers-came/)) token giveaways and learned interesting lessons about proving unique humanity & [sybil resistance](https://en.wikipedia.org/wiki/Sybil_attack)

Stuff that's interesting about Soroban, their new smart contract platform:

- uses Wasm

- focused on Rust

- creator of Rust on the team??

   <img width="1294" alt="graydon hoare created rust" src="https://user-images.githubusercontent.com/221614/193350533-87b8d0fc-13b9-47db-a28a-50245eecdbfa.png">
   
   (also, yes, that's a screenshot of [an issue](https://github.com/stellar/rs-soroban-sdk/issues/683) where they're considering switching to WIT, inspired by my team's work on [RAEN](https://raen.dev) 😊)

- contract interface / ABI added to the Wasm custom section, [RAEN](https://raen.dev/)-style, by default, so that Every. Single. Contract. will have fully-typed interfaces discoverable by all developers on the platform, right out the gate.

- event system baked into the protocol, also right out the gate (NEAR could have used this!)

- sometimes being late to the game is an asset: can learn from existing Rust & Wasm blockchains, taking what works and tweaking what doesn't

- those contract sizes!

So I dunno, Soroban & Stellar don't seem doomed.

**Comment by rice2000** [comment]
Thanks for this, @chadoh!  It's pretty eye-opening!  We'll be reviewing submissions and offering rewards starting next week, and I'll ping you here with more info once we do.  In the meantime, let me know if you have any questions.  I definitely encourage you to submit again if you're interested as well.  The more feedback and insight we can get at this stage, the better.

**Comment by gracietti** [comment]
Interesting!! I'm wondering what the comparison would look like with a more complex contract.

**Reply by chadoh** [reply]
I did a follow-up post to look into this! #26

**Comment by kalepail** [comment]
Great write up! I too would be curious how this stacks up for a more complex example. I would imagine though this is largely due to the standard library not being excluded on the Near side? Is it possible to do that or do things break? For more complex contracts would the exclusion of the standard library eventually come back to bite you as you pull in standard libraries piece meal?

I guess the real question is how significant is this comparison when expanded out to more common contract use cases?

Either way it's great to have you and thanks for a great submission!

**Reply by chadoh** [reply]
Check out the non-trivial comparison at #26

**Comment by achallagundla** [comment]
Heya Chad,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 1500 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Great piece BTW! Can't wait to see you keep tinkering with Soroban and documenting your finding! 

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya

**Comment by chadoh** [comment]
You know, 99KB to 477B isn't a fair comparison. That NEAR contract was completely unoptimized.

What's the size of the NEAR contract if we run the same optimized build settings as the Soroban contract?

[This branch of RAEN](https://github.com/raendev/raen/pull/17) adds those same optimizations to `raen build --release`. Once it's deployed, you'll be able to `cargo install raen` and then get an optimized NEAR contract with just `raen build --release`. The result: **64KB**!

<img width="1036" alt="Screen Shot 2022-10-21 at 12 33 39 PM" src="https://user-images.githubusercontent.com/221614/197245275-f0d8c355-00cf-45b5-ad3f-ef5f1f05cc43.png">

477 bytes to 64 kilobytes is still more than a 100x size difference, though! Why are NEAR contracts so big? More on that in #26!
